test/fnn/test_fnn_cpu.cpp

```diff
@@ -0,0 +1,196 @@
+
+    // Cheks if the data directory exists
+    if (!create_directory_if_not_exists(data_dir)) {
+        std::cout << "Error: could not create data directory" << std::endl;
+        return false;
+    }
+
+
+    // Train data
+    Dataloader train_db = train_data(data, tagi_net, data_path, NORMALIZE);
+    // Test data
+    Dataloader test_db =
+        test_data(data, tagi_net, data_path, train_db, NORMALIZE);
+
+    std::vector<std::vector<float> *> weights;
+    weights.push_back(&tagi_net.theta.mw);
+    weights.push_back(&tagi_net.theta.Sw);
+    weights.push_back(&tagi_net.theta.mb);
+    weights.push_back(&tagi_net.theta.Sb);
+
+    std::vector<std::vector<float> *> bias;
+    bias.push_back(&tagi_net.theta.mw_sc);
+    bias.push_back(&tagi_net.theta.Sw_sc);
+    bias.push_back(&tagi_net.theta.mb_sc);
+    bias.push_back(&tagi_net.theta.Sb_sc);
+
+    if (recompute_outputs) {
+        write_vector_to_csv(init_param_path_w, "mw,Sw,mb,Sb", weights);
+        write_vector_to_csv(init_param_path_b, "mw_sc,Sw_sc,mb_sc,Sb_sc", bias);
+    }
+
+    // Read the initial parameters (see tes_utils.cpp for more details)
+    read_vector_from_csv(init_param_path_w, weights);
+    read_vector_from_csv(init_param_path_b, bias);
+
+    // Train the network
+    regression_train(tagi_net, train_db, EPOCHS);
+
+    // Test the network
+    regression_test(tagi_net, test_db);
+
+    std::vector<std::vector<float> *> forward_states;
+    forward_states.push_back(&tagi_net.state.mz);
+    forward_states.push_back(&tagi_net.state.Sz);
+    forward_states.push_back(&tagi_net.state.ma);
+    forward_states.push_back(&tagi_net.state.Sa);
+    forward_states.push_back(&tagi_net.state.J);
+
+    std::vector<std::vector<float>> backward_states;
+    std::string backward_states_header = "mean_i,sigma_i";
+
+    for (int i = 0; i < net.layers.size() - 2; i++) {
+        backward_states.push_back(
+            std::get<0>(tagi_net.get_inovation_mean_var(i)));
+        backward_states.push_back(
+            std::get<1>(tagi_net.get_inovation_mean_var(i)));
+    }
+
+    std::vector<std::vector<float> *> backward_states_ptr;
+    for (int i = 0; i < backward_states.size(); i++)
+        backward_states_ptr.push_back(&backward_states[i]);
+
+    if (recompute_outputs) {
+        // RESET OUPUTS
+
+        // Write the parameters and hidden states
+        write_vector_to_csv(opt_param_path_w, "mw,Sw,mb,Sb", weights);
+        write_vector_to_csv(opt_param_path_b, "mw_sc,Sw_sc,mb_sc,Sb_sc", bias);
+
+        // Write the forward hidden states
+        write_vector_to_csv(forward_states_path, "mz,Sz,ma,Sa,J",
+                            forward_states);
+
+        // Write the backward hidden states
+        write_vector_to_csv(backward_states_path, backward_states_header,
+                            backward_states_ptr);
+
+    } else {
+        // PERFORM TESTS
+
+        // Read the saved parameters
+        std::vector<std::vector<float> *> updated_weights;
```

Comment:
I would call it `ref_weights` instead of `updated_weights` for the readability purpose. The same goes for bias